pkg/controllers/applicationset_controller.go

```diff
@@ -49,14 +54,24 @@ func (r *ApplicationSetReconciler) Reconcile(req ctrl.Request) (ctrl.Result, err
 		return ctrl.Result{}, client.IgnoreNotFound(err)
 	}
 
-	var generator generators.Generator
-	generator = generators.NewListGenerator()
+	listGenerator := generators.NewListGenerator()
+	clusterGenerator := generators.NewClusterGenerator(r.Client)
+
+	// desiredApplications is the main list of all expected Applications from all generators in this appset.
+	var desiredApplications  []argov1alpha1.Application
 	for _, tmpGenerator := range applicationSetInfo.Spec.Generators {
-		desiredApplications, err := generator.GenerateApplications(&tmpGenerator, &applicationSetInfo)
-		log.Infof("desiredApplications %+v", desiredApplications)
+		var apps []argov1alpha1.Application
```

Comment:
If we have many kinds of generator,  will this increase many `if else` struct .

should we provide a abstract function to hide these like factory.

Reply:
Per slack it sounds like perhaps we should instantiate all generators we know of, then call each with the AppSet, and let it return us the list of Applications for every item in the list that matches.

Although we should probably discuss if it makes sense to have multiple generators in one AppSet at some point, I'm a little worried that'll never be used due to the nature of the templating and the available params being different.

Should we defer this and followup PR once we know what to do?

Reply:
yes.  fine